arwen/elrondapi/elrondei.go

```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
 	return int32(len(esdtData.Value.Bytes()))
 }
 
+//export v1_4_getESDTLocalRoles
+func v1_4_getESDTLocalRoles(context unsafe.Pointer, tokenIdHandle int32) int64 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	storage := arwen.GetStorageContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	tokenID, err := managedType.GetBytes(tokenIdHandle)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return -1
+	}
+	key := []byte("ELRONDroleesdt" + string(tokenID))
+
+	data_buffer, usedCache := storage.GetStorage(key)
+	storage.UseGasForStorageLoad(
+		storageLoadName,
+		metering.GasSchedule().ElrondAPICost.StorageLoad,
+		len(data_buffer), usedCache)
+
+	result := int64(0)
+	current_index := 0
+	value_len := len(data_buffer)
+
+	for current_index < value_len {
+		// first character before each role is a \n, so we skip it
+		current_index += 1
+
+		// next is the length of the role as string
+		role_len := int(data_buffer[current_index])
+		current_index += 1
+
+		// next is role's ASCII string representation
+		end_index := current_index + role_len
+		role_name := data_buffer[current_index:end_index]
+		current_index = end_index
+
+		result |= 1 << binary.LittleEndian.Uint64(role_name)
+	}
+
+	return result
+}
+
+func validateToken(tokenID []byte) int32 {
+	tokenIDLen := len(tokenID)
+
+	if tokenIDLen < identifierMinLength || tokenIDLen > identifierMaxLength {
+		return 0
+	}
+
+	// ticker must be all uppercase alphanumeric
+	tickerLen := tokenIDLen - additionalRandomCharsLength
+
+	for i := 0; i < tickerLen-1; i++ {
```

Comment:
func isTickerValid(tickerName []byte) bool {
	if len(tickerName) < minLengthForTickerName || len(tickerName) > maxLengthForTickerName {
		return false
	}

	for _, ch := range tickerName {
		isBigCharacter := ch >= 'A' && ch <= 'Z'
		isNumber := ch >= '0' && ch <= '9'
		isReadable := isBigCharacter || isNumber
		if !isReadable {
			return false
		}
	}

	return true
}